mongo-store.js

```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
       } else return cb()
     },
 
-    save: function (args, cb) {
-      var ent = args.ent
+    save: function (msg, done) {
+      return getcoll(msg, msg.ent, function (err, coll) {
+        if (error(msg, err, done)) {
+          return
+        }
 
-      var update = !!ent.id
+        const is_update = null != msg.ent.id
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+        if (is_update) {
+          return updateExisting(msg, coll, done)
+        }
+
+        return createAndSave(msg, coll, done)
+      })
+
+
+      function createAndSave(msg, coll, done) {
+        if (isUpsertRequested(msg)) {
+          return doUpsert(msg, coll, done)
+        }
+
+        return createNew(msg, coll, done)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+        function isUpsertRequested(msg) {
+          if (null == msg.q) {
+            return false
           }
 
-          entp._id = makeid(id)
+          const public_entdata = msg.ent.data$(false)
+          const upsert_on = msg.q.upsert$
 
-          coll.insertOne(entp, function (err, inserts) {
-            if (!error(args, err, cb)) {
-              var entu = inserts.ops[0]
-              var fent = null
-              if (entu) {
-                entu.id = idstr(entu._id)
-                delete entu._id
-                //fent = ent.make$(_.cloneDeep(entu))
-                fent = ent.make$(seneca.util.deep(entu))
+          return Array.isArray(upsert_on) &&
+            upsert_on.length > 0 &&
+            upsert_on.every(p => p in public_entdata)
+        }
+
+        function doUpsert(msg, coll, done) {
+          const upsert_on = msg.q.upsert$
+          const public_entdata = msg.ent.data$(false)
+
+          const filter_by = upsert_on
+            .reduce((acc, field) => {
+              acc[field] = msg.ent[field]
+              return acc
+            }, {})
+
+          const replacement = (() => {
+            const o = Dot.flatten(Object.assign({}, public_entdata))
```

Comment:
We need to remove the dependency on Dot.
Introducing a dependency is a big decision that has consequences for maintenance and security and needs to be examined separately.
Also, we cannot allow the code base to become inconsistent - if Dot is used, then it has to be used in all places where it would make sense.
The introduction of something like Dot would also have to be a separate PR.

Reply:
The Dot has been listed as a dependency prior to this PR.